Add API documentation and restructure docs

jhobz opened the pull request:
Follows [the Diátaxis method](https://diataxis.fr/) of technical documentation, using [GitHub's example repository](https://github.com/github/internal-documentation-example/tree/main?tab=readme-ov-file) as inspiration.

jhobz committed:
Add API documentation and restructure docs

jhobz committed:
Tweak some formatting and ordering

jhobz committed:
Add linting for Markdown files

jhobz committed:
Add documentation for World class and list of worlds

jhobz committed:
Add table of contents to kh2lib.md

jhobz committed:
Add documentation for Room

Also include some minor styling updates.
Tables are still WIP (going to build a script for them).

jhobz committed:
Add defaults for required Event object parameters

jhobz committed:
Add documentation for events

Also add reference to hexadecimal notation across docs.

jhobz committed:
docs: Add basic README for constants directory

jhobz committed:
docs: Update constructor documentation to include properties

jhobz committed:
Fix error in 100 Acre Woods lookup

jhobz committed:
docs: [WIP] lookup tables docs

RebelWatt14 committed:
[WIP] add script for generating constant markdown tables

jhobz committed:
Move docs build script and fix module import paths

Adjust some module imports to use a relative path syntax in order to
avoid linting errors.

jhobz committed:
Change event and room lists to 1-dimensional arrays

There's really no reason these need to be split by worlds, and making
them simple arrays allows for better refactoring of other code.

jhobz committed:
Add documentation for KH2Lib module and globals

Also fix some documentation and move offsets into a private field

jhobz committed:
docs: Add explanations for some decisions

jhobz committed:
docs: Add some initial how-to guides

jhobz committed:
docs: Add KH2ValueConstant reference

jhobz committed:
docs: Add placeholder offsets reference

jhobz committed:
Add event & event_name to kh2lib.current

jhobz committed:
docs: Styling and wording updates

jhobz committed:
Fix doc for game state table

jhobz committed:
docs: Add full tutorial for writing first mod

jhobz committed:
docs: minor wording changes

jhobz committed:
docs: Move the last of the old DeveloperDocs file elsewhere

Update README to include new documentation folders

jhobz committed:
docs: styling

jhobz committed:
Remove globals from example script and update code style

Globals should be avoided in favor of locals whenever possible.

jhobz committed:
docs: minor tutorial tweaks

jhobz committed:
docs: Fix incorrect shortcut

jhobz committed:
Make build script for docs reference tables

jhobz committed:
Use locals in LuaLibrary.lua

jhobz committed:
Remove test mod

As useful as it is during development, the test mod would get included
when the library is installed with OpenKH, which would result in it
showing up in the mod list for end users.

jhobz committed:
Reorganize some files a bit

Also add blank CONTRIBUTING.md to fix a broken link in one of the docs

paulblessing commented on docs/explanations/private_properties.md:
```diff
@@ -0,0 +1,116 @@
+
+Unfortunately, the original author(s) of the Lua engines for OpenKH did not follow standard naming
+conventions. Per the [Lua 5.1 Reference Manual][1],
+> "As a convention, names starting with an underscore followed by uppercase letters (such as _VERSION)
```
Think this is more talking about underscores with `_ALL_CAPS` naming, as opposed to what `_OnInit` is. Don't mind that we mention this here as it still makes the point, but the example may not be fully apples to apples.

paulblessing commented on docs/how_to_guides/importing_the_library.md:
```diff
@@ -0,0 +1,88 @@
+    kh2lib_status, kh2lib = pcall(require, 'kh2lib')
+
+    if not kh2lib_status then
+        ConsolePrint("KH2-Lua-Library mod is not installed", 3)
```
I don't think `ConsolePrint` is available on emulator, or at least I was convinced of it when I wrote the implementations of the various `Log` functions. That's why the examples and existing scripts use `print` here.

jhobz replied:
Oh, interesting. I'll see if I can confirm that and, if so, either replicate the functionality with `print()` or do a non-null assertion on `ConsolePrint()` and use that if available (falling back to basic `print()`).

jhobz replied:
As per our Discord discussion, it turns out `ConsolePrint()` is indeed available on emulator.

paulblessing commented on docs/how_to_guides/importing_the_library.md:
```diff
@@ -0,0 +1,88 @@
+function _OnInit()
+    kh2libstatus, kh2lib = pcall(require, "kh2lib")
+    ...
+    CanExecute = kh2lib.CanExecute
```
Probably best to keep the examples consistent. The code snippet above uses `can_execute` vs. `CanExecute` here.

jhobz replied:
Bah, yeah I did the conversion later and must have missed that one. Good catch

paulblessing commented on docs/reference/kh2lib.md:
```diff
@@ -0,0 +1,292 @@
+For a full list of exposed addresses, see [Addresses][ref-addresses].
+
+- `kh2lib.Now` {integer} Current location data
+- `kh2lib.Sve` {integer} Saved location
```
I'd recommend removing this one from the docs as `Save` is far more common and having both may introduce confusion earlier than necessary.

paulblessing commented on docs/tutorials/01_creating_your_first_mod.md:
````diff
@@ -0,0 +1,470 @@
+
+    ```lua
+    if not kh2lib_status then
+        ConsolePrint("KH2-Lua-Library mod is not installed", 3)
````
Similar here as mentioned elsewhere with pretty sure `ConsolePrint` is not available on emulator.

paulblessing commented on docs/tutorials/01_creating_your_first_mod.md:
```diff
@@ -0,0 +1,470 @@
+    kh2lib_status, kh2lib = pcall(require, 'kh2lib')
+
+    if not kh2lib_status then
+        ConsolePrint("KH2-Lua-Library mod is not installed", 3)
```
and here as well

paulblessing commented on io_packages/kh2lib/constants/rooms.lua:
```diff
@@ -1,4 +1,5 @@
-local KH2ValueConstant = require('kh2lib.constants.base_class')
+local BASE_PATH = (...):match('(.-)[^%.]+$')
```
Not sure I'm understanding what's going on here, nor why it's needed.

jhobz replied:
Basically, this allows the use of relative imports. When I added the doc building script, I was receiving linting errors trying to import the same module by two different names (because the path to it was different from their respective folders). This allows their use from either (any) script. I'll add a comment.

paulblessing replied:
Feel like _something_ is necessary - even for library authors who will be the ones maintaining this code, this feels cryptic.

If this is _solely_ for the convenience of the script you're running to build the docs, I'm not quite convinced it's worth the loss of readability for those perusing the library code itself. That said, I'm not using any linting or anything from an IDE outside of just syntax highlighting, so it's possible my concerns come from a different perspective. Certainly not saying I'm fully rejecting this change, just not a huge fan so far at least.

jhobz replied:
I think a comment should likely suffice. It would be weird to include all this work for linting/language servers and just have to add a comment to disable them when you need to import the packages from elsewhere. It also makes these imports more resilient against future changes to the directory structure.

paulblessing commented on templates/kh2_lua_library_example_commented.lua:
```diff
@@ -0,0 +1,48 @@
+function _OnInit()                                              -- Runs during game initialization, only once
+    kh2lib_status, kh2lib = pcall(require, 'kh2lib')            -- Imports the KH2 Lua Library, detects game version, and populates the `kh2lib` table
+    if not kh2lib_status then                                   -- Checks if there was an error loading the library
+        ConsolePrint('KH2-Lua-Library mod is not installed', 3) -- Try to report as meaningful of an error as possible
```
...and here

paulblessing commented on templates/kh2_lua_library_example_template.lua:
```diff
@@ -0,0 +1,33 @@
+function _OnInit()
+    kh2lib_status, kh2lib = pcall(require, 'kh2lib')
+    if not kh2lib_status then
+        ConsolePrint('KH2-Lua-Library mod is not installed', 3)
```
...and here

paulblessing commented on types/LuaBackend.lua:
```diff
@@ -0,0 +1,498 @@
+
+-- #endregion
+
+-- #region | Discord RPC
```
This stuff even work that you know? I'd just as soon not document any of this Discord stuff as IMO it shouldn't even be part of the API.

jhobz replied:
I don't know for sure if it works or not, but if it's included as part of LuaBackend, I'd still rather include it here. One of the tutorials I'd like to add involves setting up your dev environment (and technically the first tutorial does recommend the linter), so if someone's using the codebase to learn how to write a mod and wants to play around with Discord RPC functionality, I'd like them to still get linting support. I can see about maybe checking if the RPC stuff works, but even if it doesn't, I think it might still make the most sense to match the LuaBackend documentation and simply forward any inquiries about it up the chain, since it's mentioned as something that _should_ work.

paulblessing replied:
I just really really wish this Discord stuff wasn't baked in as part of the core APIs of LuaBackend/Engine itself. Always felt very out of place and something that would've been far better served as some separate Discord-specific add-on. So I guess I'm kind of feeling like adding documentation for that is effectively endorsing that pattern, which leaves me feeling a bit yucky. But that said, I'm trying to do my best not to gatekeep, just wish we didn't have to make the choice at all.

jhobz replied:
While I agree in principle, I'm not sure it exactly _could_ have been added as an add-on, given that LuaBackend uses a specially-named file (`DBGHELP.dll`). And maintaining two versions of LuaBackend (e.g. LuaBackend and LuaBackend-With-Discord-RPC) just seems untenable. This documentation is only for the linter, so people are really only likely to run into it if they start typing out the relevant functions anyways. I would prefer to leave it in order to keep parity with the LuaBackend docs.

paulblessing reviewed (changes requested):
Gave a quick look. Would ultimately like to see some/all of the TODOs/WIPs resolved before considering 3.0 finalized - that said, the dev branch probably needs to live on for some amount of time while we test the actual library functionality and all, so I'm OK with merging to the dev branch even while the TODOs/WIPs are fleshed out.

jhobz committed:
Address PR comments, and switch print statements to ConsolePrint

It turns out ConsolePrint() is available on emulator, so we should just
use that everywhere instead of print().

jhobz committed:
Prefix script name with `F266B00B` when built by OpenKH

This has no effect on PC and fixes emulator compatibility, which was
previously broken, unless you knew to rename the script.

jhobz commented:
Addressed all comments. Left discussions open that were more centered around preferences, for you to resolve (or comment further).

Also now closes #7 and addresses second half of #8.

As for the WIP documentation, those that exist are not so much WIP because of not being written, but rather that the feature has not been fully fleshed out yet. For example, the "addresses" section is not a comprehensive feature yet. The current "feature" (behind a private property, `kh2lib._constants._offsets`) is intended to hint at the structural direction all of these memory addresses should move towards. But I don't actually know how most of these addresses work yet, so I wouldn't be able to properly document them. I similarly included the blank or WIP documentation as an indicator as to how (in a structural sense) they should likely be documented in the future. I can remove the files, if desired, but I don't think filling them out right now makes much sense, unless accompanied by a full inclusion of more of the addresses/offsets. Unsure if that all makes sense or not, but that's where my thinking was with those documents.

jhobz committed:
Fix error when world ID is 0 (which happens on emulator bootup)

Note: a more resilient fix should be put in place to detect *any*
unknown world or room IDs

jhobz committed:
docs: Fix some minor issues and add params to global functions
